### PR TITLE
v1: compatibility breaking change - support native eth_address

### PR DESCRIPTION
**Pull request by Flouse**
## What's Changed
* fix erc20 total supply by @zeroqn in https://github.com/nervosnetwork/godwoken-polyjuice/pull/115
* Fix memory leaks in bn256_pairing_istanbul() by @Flouse in https://github.com/nervosnetwork/godwoken-polyjuice/pull/99
* Patch generator to support AOT VM mode by default by @Flouse in https://github.com/nervosnetwork/godwoken-polyjuice/pull/125
* BREAKING CHANGE: support native eth_address by the ETH Address Registry layer2 builtin contract by @magicalne & @Flouse in https://github.com/nervosnetwork/godwoken-polyjuice/pull/135

### CODE REVIEW

File: c/polyjuice_utils.h [hunk 1]
```diff
@@ -38,28 +32,58 @@ void debug_print_data(const char *prefix, const uint8_t *data,
...
-static const uint32_t CONTRACT_ACCOUNT_SCRIPT_ARGS_SIZE = 32 + 4 + 20;
+/* https://stackoverflow.com/a/1545079 */
+#pragma push_macro("errno")
+#undef errno
+bool is_errno_ok(mol_seg_res_t *script_res) {
+  return script_res->errno == MOL_OK;
+}
+#pragma pop_macro("errno")
+
+gw_reg_addr_t init_reg_addr(const uint8_t eth_addr[ETH_ADDRESS_LEN]) {
```
Comment (TheWaWaR) [inline c/polyjuice_utils.h:68]: If return the `gw_reg_addr_t` from current stack memory, better use `new_reg_addr` than `init_reg_addr` as function name.

File: c/sudt_contracts.h [hunk 2]
```diff
@@ -161,11 +208,20 @@ int transfer_to_any_sudt(gw_context_t* ctx,
     return ERROR_TRANSFER_TO_ANY_SUDT;
   }
 
-  evmc_address from_address = *((evmc_address *)(input_src + 32 + 12));
-  evmc_address to_address = *((evmc_address *)(input_src + 64 + 12));
-  ret = sudt_transfer(ctx, sudt_id, POLYJUICE_SHORT_ADDR_LEN, from_address.bytes, to_address.bytes, amount);
+  gw_reg_addr_t from_addr = {0};
+  from_addr.reg_id = GW_DEFAULT_ETH_REGISTRY_ACCOUNT_ID;
+  memcpy(from_addr.addr, input_src + 32 + 12, ETH_ADDRESS_LEN);
+  from_addr.addr_len = ETH_ADDRESS_LEN;
```
Comment (TheWaWaR) [inline c/sudt_contracts.h:214]: use function to build `gw_reg_addr_t` value.

File: c/sudt_contracts.h [hunk 3]
```diff
@@ -161,11 +208,20 @@ int transfer_to_any_sudt(gw_context_t* ctx,
...
+
+  gw_reg_addr_t to_addr = {0};
+  to_addr.reg_id = GW_DEFAULT_ETH_REGISTRY_ACCOUNT_ID;
+  memcpy(to_addr.addr, input_src + 64 + 12, ETH_ADDRESS_LEN);
+  to_addr.addr_len = ETH_ADDRESS_LEN;
```
Comment (TheWaWaR) [inline c/sudt_contracts.h:219]: use function to build `gw_reg_addr_t` value.

Comment (TheWaWaR) [reply c/sudt_contracts.h:219]: Please search all usage of `GW_DEFAULT_ETH_REGISTRY_ACCOUNT_ID`, and replace the code to function call.

File: c/sudt_contracts.h (hunk 2 above)
Comment (Flouse) [reply c/sudt_contracts.h]: https://github.com/nervosnetwork/godwoken-polyjuice/pull/138/commits/0ea8e4e4b8ded3dfb992293af381f3699f495900

File: c/polyjuice_utils.h (hunk 1 above)
Comment (Flouse) [reply c/polyjuice_utils.h]: https://github.com/nervosnetwork/godwoken-polyjuice/pull/138/commits/0ea8e4e4b8ded3dfb992293af381f3699f495900

File: c/sudt_contracts.h (hunk 3 above)
Comment (Flouse) [reply c/sudt_contracts.h]: https://github.com/nervosnetwork/godwoken-polyjuice/pull/138/commits/0ea8e4e4b8ded3dfb992293af381f3699f495900

File: .gitmodules [hunk 4]
```diff
@@ -27,4 +27,5 @@
 	url = https://github.com/mohanson/bn128.git
 [submodule "deps/godwoken-scripts"]
 	path = deps/godwoken-scripts
-	url = https://github.com/nervosnetwork/godwoken-scripts.git
+	url = https://github.com/jjyr/godwoken-scripts.git
```
Comment (jjyr) [inline .gitmodules:30]: Please use master branch, the [PR](https://github.com/nervosnetwork/godwoken-scripts/pull/116) is merged

Comment (Flouse) [reply .gitmodules]: done

File: c/polyjuice.h [hunk 5]
```diff
@@ -299,36 +283,37 @@ int load_account_code(gw_context_t* gw_ctx, uint32_t account_id,
...
-  if (ret != 0) {
-    debug_print_int("get script hash by block producer id failed", context->gw_ctx->block_info.block_producer_id);
-    context->error_code = ret;
-  }
-  memcpy(ctx.block_coinbase.bytes, coinbase_script_hash, 20);
+
+  gw_reg_addr_t *block_producer = &context->gw_ctx->block_info.block_producer;
+  debug_print_data("load block_coinbase address:",
+                   block_producer->addr, ETH_ADDRESS_LEN);
+  memcpy(ctx.block_coinbase.bytes, block_producer->addr, ETH_ADDRESS_LEN);
```
Comment (jjyr) [inline c/polyjuice.h:315]: The length of block_producer->addr is not fixed to 20 bytes, we should check the length first, and return an error if it is invalid.

**Review by magicalne — DISMISSED** [review]

File: c/polyjuice.h [hunk 6]
```diff
@@ -949,25 +1013,20 @@ int handle_message(gw_context_t* ctx,
...
-  // TODO: passing this value from function argument
-  uint8_t from_script_hash[32] = {0};
-  uint8_t to_script_hash[32] = {0};
+
   if (memcmp(zero_address.bytes, msg.destination.bytes, 20) != 0) {
-    ret = ctx->sys_get_script_hash_by_prefix(ctx, msg.destination.bytes, 20, to_script_hash);
-    if (ret == 0) {
+    ret = load_account_id_by_eth_address(ctx, msg.destination.bytes, &to_id);
+    if (ret != 0) {
+      debug_print_int(
```
Comment (jjyr) [inline c/polyjuice.h:1026]: Shouldn't we return error here? Miss the `return ret;`

File: c/polyjuice_globals.h [hunk 7]
```diff
@@ -1,18 +1,40 @@
...
+
+#define ETH_ADDRESS_LEN 20
+
+/* Key type for ETH Address Registry */
+#define GW_ACCOUNT_SCRIPT_HASH_TO_ETH_ADDR 200
+#define ETH_ADDR_TO_GW_ACCOUNT_SCRIPT_HASH 201
+
+/** Polyjuice contract account (normal/create2) script args size */
+#define CONTRACT_ACCOUNT_SCRIPT_ARGS_LEN 56 /* 32 + 4 + 20 */
+#define CREATOR_SCRIPT_ARGS_LEN 40			/* 32 + 4 + 4  */
```
Comment (jjyr) [inline c/polyjuice_globals.h:14]: Should be 36.

32 bytes rollup scripts hash | 4 bytes sudt_id

the registry id is fixed to ETH_REGISTRY_ACCOUNT_ID.

**Review by jjyr — CHANGES_REQUESTED** [review]

File: README.md [hunk 8]
```diff
@@ -72,33 +74,30 @@ info_content:
...
-
-### Address used in polyjuice
-
-The address used in polyjuice are all godwoken short address, which is:
+### Address used in Polyjuice
+In the latest version of Polyjuice, the [EOA](https://ethereum.org/en/glossary/#eoa) address is native `eth_address`, which is the rightmost 160 bits of a Keccak hash of an ECDSA public key. While the address for an Polyjuice contract is deterministically computed from the address of its creator (sender) and how many transactions the creator has sent (nonce). The sender and nonce are RLP encoded and then hashed with Keccak-256.
 
-``` rust
-short_address = blake2b(script.as_slice())[0..20]
+> In the previous version of Polyjuice, all the addresses are `short_godwoken_account_script_hash`, which is:
```
[the remark this answers is not in the document]
Comment (Flouse) [reply README.md]: https://github.com/nervosnetwork/godwoken-polyjuice/pull/138/commits/9ba67f391c9699356791649f41a333d936f7289c

File: c/polyjuice.h (hunk 5 above)
Comment (Flouse) [reply c/polyjuice.h]: https://github.com/nervosnetwork/godwoken-polyjuice/pull/138/commits/24cc37b1ea5687ae156351104006878c4ce2e018

File: c/polyjuice_globals.h (hunk 7 above)
Comment (Flouse) [reply c/polyjuice_globals.h:14]: `gw-tools create_creator_account` created a creator_account with CREATOR_SCRIPT_ARGS_LEN = 40.

see:
https://github.com/nervosnetwork/godwoken/blob/982c879/crates/tools/src/create_creator_account.rs#L57-L59

File: c/polyjuice.h (hunk 6 above)
Comment (Flouse) [reply c/polyjuice.h:1034]: This `if` branch only indicates that `to_address_exists = false`, we should not return ret here.

Scenario example: Call an ERC20 contract to transfer tokens to an eth_address which related Godwoken account was not created.

File: Cargo.toml [hunk 9]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "godwoken-polyjuice"
-version = "0.6.0"
+version = "1.0.0"
 authors = ["Linfeng Qian <thewawar@gmail.com>"]
```
[the remark this answers is not in the document]
Comment (Flouse) [reply Cargo.toml]: 0964c275f59d35876d507a3a7b046bbe3602ac3a

File: c/polyjuice_globals.h (hunk 7 above)
Comment (Flouse) [reply c/polyjuice_globals.h]: 0d94b46d10dfc36f12e65154dfc72cb68aac28b7

**Review by jjyr — APPROVED** [review]

**Review by TheWaWaR — APPROVED** [review]

**Pull request merged**